base: add `Sys.detectwsl()`

inkydragon opened the pull request:
Close #36425, fix #36354


## How to detect WSL?

There are a number of ways that can be used to detect WSL environments,
but each can have false positives.

We finally chose to use the same method as Snapd to detect WSL.
Because Windows installs Ubuntu LTS as WSL by default.
So we assume that Snapd's detection method will work for most users.

- Ubuntu/Snapd: https://github.com/canonical/snapd/blob/03a578a5dff26467dcc80580fcd4720a486185a5/release/release.go#L151-L172
- microsoft/WSL#423
- microsoft/WSL#4071
- https://superuser.com/q/1749781/1460597


## Known limitations

- this is a runtime test, and thus cannot meaningfully be used in `@static if` constructs.
- Linux users can create their own
	- `/proc/sys/fs/binfmt_misc/WSLInterop` file
	- or `/run/WSL/` folder

	to pretend to be a WSL environment.


--- 

- I've tested this under: Ubuntu 22.04.5 LTS (default/Offical) and alpine-release-3.17.0 (win store)
- [x] Add compat, NEWS, tests
- [x] Take a look at different detect methods,
  figure out which one is more robust

mgautam98 committed:
detectwsl

inkydragon committed:
Apply suggestion

Co-Authored-By: Max Horn <241512+fingolfin@users.noreply.github.com>

inkydragon committed:
base.detectwsl: update compat to v1.12

inkydragon committed:
test.path: use `Sys.detectwsl`

inkydragon committed:
Add to NEWS

inkydragon commented:
```
error during bootstrap:
LoadError("sysimg.jl", 5, LoadError("Base.jl", 127, LoadError("sysinfo.jl", 3, UndefVarError(:isfile, Base.Sys))))
ijl_undefined_var_error at [buildroot]/src/rtutils.c:152
detectwsl at ./sysinfo.jl:550
```

Maybe I need to use `Base.Filesystem.isfile`.
But why this line works?
https://github.com/JuliaLang/julia/blob/316f0fe0b0e454093b44095bb99b273d4158bdc0/base/sysinfo.jl#L619

ararslan commented on base/sysinfo.jl:
```diff
@@ -532,7 +533,26 @@ including e.g. a WebAssembly JavaScript embedding in a web browser.
 
-for f in (:isunix, :islinux, :isbsd, :isapple, :iswindows, :isfreebsd, :isopenbsd, :isnetbsd, :isdragonfly, :isjsvm)
+"""
+    Sys.detectwsl([os])
```
Perhaps `Sys.iswsl()` for consistency with the other platform detection functions?

inkydragon replied:
This is actually on purpose, as `detectwsl` detects at runtime, so names of the form `is_platform` are not used.
See discuss in #36354

ararslan commented on base/sysinfo.jl:
```diff
@@ -532,7 +533,26 @@ including e.g. a WebAssembly JavaScript embedding in a web browser.
+!!! compat "Julia 1.12"
+    This function requires at least Julia 1.12.
+"""
+function detectwsl(os::Symbol)
+    islinux(os) &&
```
Since this function can only determine whether the current system is WSL, it can't correctly query based on an arbitrary kernel, I don't think it makes sense to include the `os` argument.
```suggestion
function detectwsl()
    islinux() &&
```

ararslan commented on base/sysinfo.jl:
```diff
@@ -532,7 +533,26 @@ including e.g. a WebAssembly JavaScript embedding in a web browser.
+    contains(read("/proc/sys/kernel/osrelease", String), r"Microsoft|WSL"i)
+end
+
+for f in (:detectwsl, :isunix, :islinux, :isbsd, :isapple, :iswindows, :isfreebsd, :isopenbsd, :isnetbsd, :isdragonfly, :isjsvm)
```
```suggestion
for f in (:isunix, :islinux, :isbsd, :isapple, :iswindows, :isfreebsd, :isopenbsd, :isnetbsd, :isdragonfly, :isjsvm)
```
This should fix the error during building. As you noted, `isfile` is used elsewhere in this file, but the difference is that here we're attempting to actually call `detectwsl(KERNEL)` inside of this loop but can't because this file is included in the build before the one that defines `isfile`. (It's also not correct to have it in the loop here anyway since, as you note in the docstring, it's a runtime check that can't be statically evaluated.)

davidanthoff commented:
There is already a WSL detection in base at https://github.com/JuliaLang/julia/blob/316f0fe0b0e454093b44095bb99b273d4158bdc0/base/path.jl#L658 and it uses a slightly different approach. Would be good to figure out which one is more robust and then use that one consistently everywhere.

ararslan commented:
As a data point, Ubuntu's Snap daemon looks for `/proc/sys/fs/binfmt_misc/WSLInterop` to detect whether it's running on WSL, but it's apparently also possible to override the name to something other than `WSLInterop`. Looks like the approach implemented here is also fairly common.

Some references:
- https://github.com/microsoft/WSL/issues/423
- https://github.com/microsoft/WSL/issues/4071
- https://superuser.com/a/1749811

inkydragon committed:
Apply suggestions from code review

Co-authored-by: Alex Arslan <ararslan@comcast.net>

inkydragon commented on test/osutils.jl:
```diff
@@ -29,6 +29,11 @@ using Libdl
         @test Sys.windows_version() >= v"1.0.0-"
     end
+
+    @test !Sys.detectwsl(:Windows)
```
```suggestion
    @test !Sys.detectwsl()
```

inkydragon commented on base/sysinfo.jl:
```diff
@@ -532,6 +533,25 @@ including e.g. a WebAssembly JavaScript embedding in a web browser.
 isjsvm(os::Symbol) = (os === :Emscripten)
 
+"""
+    Sys.detectwsl([os])
```
```suggestion
    Sys.detectwsl()
```

inkydragon committed:
Update test/osutils.jl

inkydragon committed:
Update base/sysinfo.jl

davidanthoff commented:
I read through the links that @ararslan posted, and to me the `ispath("/proc/sys/fs/binfmt_misc/WSLInterop")` methods seems more robust and simplest. My understanding is that one can only get false positives or false negatives there if a user _really_ went out of their way to reconfigure things, essentially on purpose trying to prevent detection.

The superuser post also mentions on the other hand that the method implemented here in this PR might give a false positive maybe even on Azure, where MS might be running a custom kernel, and so greping for just "Microsoft" might get fooled.

I also think that if Ubuntu uses a given approach for snap, then it is probably a good idea to just follow that. Ubuntu and MS have worked really closely together on this whole WSL stuff, so just following their lead strikes me as a good idea.

ararslan commented:
I should note that my claim about Snap comes from the linked answer on SuperUser, which is from 2022; it's possible that Ubuntu has changed the implementation since then.

inkydragon committed:
test: remove dup test, add TODO

inkydragon committed:
base.sysinfo: use snapd ways to detect WSL

https://github.com/canonical/snapd/blob/03a578a5dff26467dcc80580fcd4720a486185a5/release/release.go#L151-L172

inkydragon commented:
> it's possible that Ubuntu has changed the implementation since then.

They added testing for the existence of “/run/WSL/” as a fallback.
We can use the same way to detect WSL.


```go
// We detect WSL via the existence of /proc/sys/fs/binfmt_misc/WSLInterop
// Under some undocumented circumstances this file may be missing. We have /run/WSL as a backup.
//
// We detect WSL1 via the root filesystem type:
// - wslfs or lxfs mean WSL1
// - Anything else means WSL2
// After knowing we're in WSL, if any error occurs we assume WSL2 as it is the more flexible version
func getWSLVersion() int {
	if !fileExists("/proc/sys/fs/binfmt_misc/WSLInterop") && !fileExists("/run/WSL") {
		return 0
	}
	fstype, err := filesystemRootType()
	if err != nil {
		// TODO log error here once logger can be imported without circular imports
		return 2
	}

	if fstype == "wslfs" || fstype == "lxfs" {
		return 1
	}
	return 2
}
```
https://github.com/canonical/snapd/blob/03a578a5dff26467dcc80580fcd4720a486185a5/release/release.go#L151-L172

inkydragon commented on base/sysinfo.jl:
```diff
@@ -532,6 +533,27 @@ including e.g. a WebAssembly JavaScript embedding in a web browser.
+Windows Subsystem for Linux (WSL).
+
+!!! note
+    Unlike `Sys.iswindows`, `Sys.islinux` etc., this is a runtime tests, and thus
+    cannot meaningfully be used in `@static if ` constructs.
```
```suggestion
    Unlike `Sys.iswindows`, `Sys.islinux` etc., this is a runtime test, and thus
    cannot meaningfully be used in `@static if` constructs.
```

inkydragon committed:
Update base/sysinfo.jl

inkydragon committed:
Merge branch 'master' into cyhan/wsl
